File: app/lib/Models/Product.dart
import 'package:skate/Models/Category.dart';

class Product {
  String _id;
  String name;
  String description;
  Category category;
  String image;
  int price;
  List<ProductStock> stock;

  String get id => _id;

  Product.fromJson(Map<String, dynamic> json) {
    if (json == null) return;
    _id = json['id'];
    name = json['name'];
    description = json['description'];
    category = Category.fromJson(json['category']);
    image = json['image'];
    price = json['price'];
    stock = json['stock']
            ?.map<ProductStock>((json) => ProductStock.fromJson(json))
            ?.toList() ??
        [];
  }

  Map<String, dynamic> toJson() => {
        "id": id,
        "name": name,
        "description": description,
        "category": category.id,
        "image": image,
        "price": price,
        "stock": stock.map((e) => e.toJson()).toList(),
      };

  Product({
    String id,
    this.name,
    this.description,
    this.category,
    this.image,
    this.price,
    this.stock,
  }) : _id = id;
}

class ProductStock {
  String size;
  int value;

  ProductStock.fromJson(Map<String, dynamic> json) {
    if (json == null) return;
    size = json['size'];
    value = json['value'];
  }

  Map<String, dynamic> toJson() => {
        "size": size,
        "value": value,
      };

  ProductStock({
    this.size,
    this.value,
  });
}

File: backend/models/product.js
var mongoose = require('mongoose');

var Product = new mongoose.Schema({
  createdDateTime: Date,
  updatedDateTime: Date,
  name: String,
  description: String,
  category: String,
  size: String,
  image: String,
  price: Number,
  stock: Array
});

Product.pre('save', function(callback) {
  var model = this;
  if (model.createdDateTime === null){
    model.createdDateTime = new Date();
  }
  model.updatedDateTime = new Date();
  callback();
});

module.exports.ProductModel = mongoose.model('product', Product);
